sdk/enoki/src/EnokiFlow.ts

```diff
@@ -0,0 +1,311 @@
+			});
+
+		const params = new URLSearchParams({
+			...input.extraParams,
+			nonce,
+			client_id: input.clientId,
+			redirect_uri: input.redirectUrl,
+			response_type: 'id_token',
+			// TODO: Eventually fetch the scopes for this client ID from the Enoki service:
+			scope: [
+				'openid',
+				// Merge the requested scopes in with the required openid scopes:
+				...(input.extraParams && 'scope' in input.extraParams
+					? (input.extraParams.scope as string[])
+					: []),
+			]
+				.filter(Boolean)
+				.join(' '),
+		});
+
+		let oauthUrl: string;
+		switch (input.provider) {
+			case 'google': {
+				oauthUrl = `https://accounts.google.com/o/oauth2/v2/auth?${params}`;
+				break;
+			}
+
+			case 'facebook': {
+				oauthUrl = `https://www.facebook.com/v17.0/dialog/oauth?${params}`;
+				break;
+			}
+
+			case 'twitch': {
+				params.set('force_verify', 'true');
+				oauthUrl = `https://id.twitch.tv/oauth2/authorize?${params}`;
+				break;
+			}
+
+			default:
+				throw new Error(`Invalid provider: ${input.provider}`);
+		}
+
+		this.$zkLoginState.set({ provider: input.provider });
+		await this.#setSession({
+			expiresAt: estimatedExpiration,
+			maxEpoch,
+			randomness,
+			ephemeralKeyPair: ephemeralKeyPair.export().privateKey,
+		});
+
+		return oauthUrl;
+	}
+
+	// TODO: Should our SDK manage this automatically in addition to exposing a method?
+	// TODO: Should we rename this? Something with "callback" maybe?
+	async handleAuthRedirect(hash: string = window.location.hash) {
+		const params = new URLSearchParams(hash.startsWith('#') ? hash.slice(1) : hash);
+
+		// Before we handle the auth redirect and get the state, we need to restore it:
+		const zkp = await this.getSession();
+
+		if (!zkp || !zkp.ephemeralKeyPair || !zkp.maxEpoch || !zkp.randomness) {
+			throw new Error(
+				'Start of sign-in flow could not be found. Ensure you have started the sign-in flow before calling this.',
+			);
+		}
+
+		const jwt = params.get('id_token');
+		if (!jwt) {
+			throw new Error('Missing ID Token');
+		}
+
+		const decodedJwt = decodeJwt(jwt);
+		if (!decodedJwt.sub || !decodedJwt.aud || typeof decodedJwt.aud !== 'string') {
+			throw new Error('Missing JWT data');
+		}
+
+		const { address, salt } = await this.#enokiClient.getZkLogin({ jwt });
+
+		this.$zkLoginState.set({
+			...this.$zkLoginState.get(),
+			salt,
+			address,
+		});
+		await this.#setSession({
+			...zkp,
+			jwt,
+		});
+
+		return params.get('state');
+	}
+
+	async #setSession(newValue: ZkLoginSession | null) {
+		if (newValue) {
+			const storedValue = await this.#encryption.encrypt(
+				this.#encryptionKey,
+				JSON.stringify(newValue),
+			);
+
+			this.#store.set(STORAGE_KEYS.SESSION, storedValue);
+		} else {
+			this.#store.delete(STORAGE_KEYS.SESSION);
+		}
+
+		this.#zkLoginSession = newValue;
+	}
+
+	async getSession() {
+		if (this.#zkLoginSessionInitialized) {
+			return this.#zkLoginSession;
+		}
```

Comment:
Not sure if this might cause issues, but the async initialization here allows trying to do this multiple times in parallel.  I think that's probably fine, but we could save the promise for the initial result, and return that on subsequent calls

Reply:
Yeah this is a good point, I just don't know if this is something I'm really worried about practically, especially given how fast this async function will be, I think the risk is pretty low